### PR TITLE
make task docstrings more verbose after being confused by interface

### PR DESCRIPTION
**Pull request by evanfwelch**
# Context
I recently started using `tfrs.recommenders` at work and for side projects. It's an amazing package with wonderful tutorials, thank you @maciejkula !

In learning the codebase, I encountered confusion about the scope of a `tfrs.task.Task`. For `Retrieval` we the task did a lot of handholding for us, expecting only the outputs of each tower, for `Ranking` we have to "bring our own scores and labels". Given the complexity of recommender architectures, I don't dispute that this is the right design but I wanted to make it clearer in the docstrings that the "interface" between application-specific model code and `tfrs.Task` changes depending on whether you're doing Ranking or Retrieval. 

# This PR
* Adds a few more explanatory sentences to the docstrings for `tfrs.tasks.Ranking` and `tfrs.tasks.Retrieval`

### CODE REVIEW

**Comment by evanfwelch** [comment]
@googlebot I fixed it.

**Comment by evanfwelch** [comment]
Hi @maciejkula, I'm a former student of Joshua Gordon (would cc but cannot find his GH handle). I hope you don't mind the PR w/out an issue creation--I didn't want to complain about the source of my confusion without just providing a suggestion for how to clarify the interface of `tfrs.tasks.*`.

File: tensorflow_recommenders/tasks/retrieval.py [hunk 1]
```diff
@@ -36,7 +36,12 @@ class Retrieval(tf.keras.layers.Layer, base.Task):
   This task defines models that facilitate efficient retrieval of candidates
   from large corpora by maintaining a two-tower, factorized structure: separate
   query and candidate representation towers, joined at the top via a lightweight
-  scoring function.
+  scoring function. Unlike tfrs.tasks.Ranking, this retrieval task assumes your
```
Comment (maciejkula) [inline tensorflow_recommenders/tasks/retrieval.py:39]: If we're expanding this, perhaps we should mention why we impose this factorized structure here? Something along the lines of:

This two-tower factorized structure is required for retrieval tasks, because it allows the candidate representations to be computed ahead of time and stored in a fast retrieval index (such as `tfrs.layers.factorized_top_k.ScaNN`). Consequently, this tasks assumes that your.... (the rest of your edit follows).

File: tensorflow_recommenders/tasks/retrieval.py [hunk 2]
```diff
@@ -109,10 +114,10 @@ def call(self,
       candidate_embeddings: [num_queries, embedding_dim] tensor of candidate
         representations.
       sample_weight: [num_queries] tensor of sample weights.
-      candidate_sampling_probability: Optional tensor of candidate sampling
+      candidate_sampling_probability: Optional [m, n] tensor of candidate sampling
```
Comment (maciejkula) [inline tensorflow_recommenders/tasks/retrieval.py:117]: Could you spell out what `m` and `n` are here?

File: tensorflow_recommenders/tasks/ranking.py [hunk 3]
```diff
@@ -34,7 +33,10 @@ class Ranking(tf.keras.layers.Layer, base.Task):
 
   This task helps with building ranker models. Usually, these will involve
   predicting signals such as clicks, cart additions, likes, ratings, and
-  purchases.
+  purchases. Accordingly, this task is less prescriptive than `tfrs.Rerieval`
```
Comment (maciejkula) [inline tensorflow_recommenders/tasks/ranking.py:36]: Typo: `tfrs.Rerieval` -> `tfrs.tasks.Retrieval`

**Review by maciejkula — CHANGES_REQUESTED** [review]
Thank you! Could you have a look at my comments?

File: tensorflow_recommenders/tasks/ranking.py (hunk 3 above)
Comment (evanfwelch) [reply tensorflow_recommenders/tasks/ranking.py:36]: > Thank you! Could you have a look at my comments?

You've been very patient, yes will address those comments right now!

File: tensorflow_recommenders/tasks/retrieval.py (hunk 1 above)
Comment (evanfwelch) [reply tensorflow_recommenders/tasks/retrieval.py:39]: Yes this is really useful context, will add and make sure it flows.

File: tensorflow_recommenders/tasks/retrieval.py (hunk 2 above)
Comment (evanfwelch) [reply tensorflow_recommenders/tasks/retrieval.py:120]: oops, yep.

**Comment by evanfwelch** [comment]
Hi @maciejkula all comments have been addressed, mind spot checking my sizes in the retrieval docstring? Thanks again! I have 2 other PRs on the way to fix two other small points of confusion

**Comment by evanfwelch** [comment]
@maciejkula regarding this PR, I will also add a changelog entry and version bump after the "typo" PR merges

**Comment by evanfwelch** [comment]
Hi @maciejkula friendly nudge on these docstring updates, no hard feelings if you don't want to merge, just let me know!
